File: momose/src/models/randomWalkModel/RandomWalkNode.java
package models.randomWalkModel;

import engine.SimTime;

import math.Point2d;
import math.Vector2d;

import models.simpleModelStructures.SimpleNode;


public class RandomWalkNode extends SimpleNode
{
    protected float localTime;
    
    public RandomWalkNode(Point2d pos, float radius) 
    {
        super(pos,radius);
        
        localTime=0;
        //Velocità iniziale
        //setVelocity(new Vector2d(0f,vMax));
    }//Fine costruttore
    
    public void think(SimTime time) 
    {
        //System.out.println("Sto pensando...");  
        //Movimento di prova
        //testMove(time);
        //Movimento random
        randomMove(time); 
    }//Fine think
    
    private void testMove(SimTime time)
    {
        //setVelocity(new Vector2d(1f,0f));  
        if(collided==true)
        {
            Vector2d tv=getVelocity();
            //setVelocity(new Vector2d((-tv.x),-(vMax)));
        } 
    }//Fine testMove
    
    private void randomMove(SimTime time)
    {
        if((localTime>=pauseTime)||(collided==true)) 
        { 
            //Calcolo il nuovo vettore velocità 
            v.x=(float)(Math.random()*vMax+vMin)*(Math.random()<0.5?(-1):(1));
            v.y=(float)(Math.random()*vMax+vMin)*(Math.random()<0.5?(-1):(1));
            
            //Rapporto la velocità all'intervallo di tempo 
            v.x=v.x*time.getDT();
            v.y=v.y*time.getDT();
            
            //System.out.println("Nuova velocità="+v);
            localTime=0;  
        } 
        
        //Incremento il tempo locale  
        localTime+=time.getDT();
        
    }//Fine randomMove
}//Fine RandomWalkNode

File: momosecpp/src/RandomWalkNode.cpp
#include "RandomWalkNode.h"

#include <stdlib.h>
#include <time.h>

RandomWalkNode::RandomWalkNode(Point2D pos,const float &radius): SimpleNode(pos,radius) 
   {
	localTime=0;
	//Imposto il seme per generare i numeri casuali
	srand((unsigned)time(NULL));
	//Velocità iniziale
	//setVelocity(Vector2D(1.0f,0.0f));
   }//Fine costruttore

RandomWalkNode::~RandomWalkNode(){}

void RandomWalkNode::think(SimTime *simTime) 
   {
    //time->dynamicTimetoXml(cout);	
	//cout<<"Sto pensando..."<<endl;  
	//Movimento di prova
	//testMove(simTime);
    //Movimento random
	randomMove(simTime);	
  }//Fine think
   
void RandomWalkNode::testMove(SimTime *simTime)
   {
	setVelocity(Vector2D(1.0f,0.0f));  
	//cout<<"Node="<<*this<<endl;
	collided=true;
	if(collided==true)
	{
	 Vector2D tv=getVelocity();
	 setVelocity(Vector2D((-tv.x),-(tv.y)));
	} 
   }//Fine testMove  
   
void RandomWalkNode::randomMove(SimTime *simTime)
   {
   	if((localTime>=pauseTime)||(collided==true)) 
	 {	
	  //Calcolo il nuovo vettore velocità	
      v.x=(float)((((float)rand()/(float)(RAND_MAX))*(vMax-vMin))+vMin)
   	                *(rand()%2<1?(-1):(1));//Segno
	  v.y=(float)((((float)rand()/(float)(RAND_MAX))*(vMax-vMin))+vMin)
   	                *(rand()%2<1?(-1):(1));//Segno
      //cout<<"v="<<v<<endl;
      
	  //Rapporto la velocità all'intervallo di tempo 
	  //v.x=v.x*simTime->getDT();
	  //v.y=v.y*simTime->getDT();
	  v*=simTime->getDT();
	  
	  //Azzero il tempo locale
	  localTime=0;  
	 }	
  
  //Incremento il tempo locale  
  localTime+=simTime->getDT();    
 }//Fine randomMove   
